File: lib/services/natural_language_normalizer.dart
import '../models/natural_language_models.dart';

/// Safe, bounded French normalization used before deterministic routing.
///
/// It deliberately preserves negation and the two meanings of "plus".
final class NaturalLanguageNormalizer {
  const NaturalLanguageNormalizer();

  static const Map<String, String> _safeTokens = {
    'prebdre': 'prendre',
    'ajoutr': 'ajouter',
    'demian': 'demain',
    'mtn': 'maintenant',
    'rapel': 'rappel',
    'rdv': 'rendez vous',
    'stp': 's il te plait',
    'aujourdhui': 'aujourd hui',
    'doeuf': 'd oeuf',
    'doeufs': 'd oeufs',
  };

  NaturalLanguageNormalization normalize(String input) {
    final codes = <String>[];
    final ambiguities = <String>[];
    var value = input.trim();
    if (value != input) codes.add('trimmed');
    final lower = value.toLowerCase();
    if (lower != value) codes.add('lowercased');
    value = lower.replaceAll(RegExp(r'[’‘`´]'), "'");
    if (value != lower) codes.add('apostrophe_normalized');

    final folded = _foldAccents(value);
    if (folded != value) codes.add('accents_folded');
    value = folded
        .replaceAll(RegExp(r"['-]"), ' ')
        .replaceAll(RegExp(r'[.!?;,:()\[\]{}"…]+'), ' ')
        .replaceAll(RegExp(r'[\u{1F300}-\u{1FAFF}]', unicode: true), ' ')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();

    final beforeOral = value;
    value = value
        .replaceFirst(RegExp(r'^jai\b'), 'j ai')
        .replaceFirst(RegExp(r'^ya\b'), 'y a')
        .replaceFirst(RegExp(r'^jveu\b'), 'je veux')
        .replaceFirst(RegExp(r'^je veu\b'), 'je veux')
        .replaceAll(RegExp(r'\bneuf heure et demie\b'), '09:30')
        .replaceAll(RegExp(r'\bneuf heures trente\b'), '09:30')
        .replaceAll(RegExp(r'\bneuf heure\b'), '09:00');
    if (value != beforeOral) codes.add('oral_form_normalized');

    final tokens = value.isEmpty ? <String>[] : value.split(' ');
    var corrected = false;
    for (var index = 0; index < tokens.length; index++) {
      final replacement = _safeTokens[tokens[index]];
      if (replacement != null) {
        tokens[index] = replacement;
        corrected = true;
      }
    }
    value = tokens.join(' ').replaceAll(RegExp(r'\s+'), ' ').trim();
    if (corrected) codes.add('safe_typo_corrected');

    if (RegExp(r'\b(?:je veux|ajoute)\s+plus\b').hasMatch(value)) {
      ambiguities.add('positive_or_quantity_plus');
    }
    if (RegExp(r'\b(?:j ai|y a|on a)\s+plus\b').hasMatch(value) &&
        !RegExp(r'\b(?:ne|n)\b').hasMatch(value)) {
      ambiguities.add('stockout_or_additional_plus');
    }
    if (RegExp(
          r'\b(?:ne|n)\s+(?:cree|creer|ajoute|ajouter|annule|annuler|'
          r'supprime|supprimer|decale|deplacer|veux)\b.{0,40}\b(?:pas|plus)\b',
        ).hasMatch(value) ||
        RegExp(
          r'\b(?:cree|ajoute|annule|supprime|decale|veux)\s+pas\b',
        ).hasMatch(value)) {
      ambiguities.add('critical_negation_scope');
    }
    if (RegExp(r'^(?:mets|met|decale)\s+(?:le|la|l)\b').hasMatch(value)) {
      ambiguities.add('unresolved_reference');
    }
    final actionCount = RegExp(
      r'\b(?:cree|ajoute|achete|annule|decale|deplace|supprime|'
      r'rappelle|planifie|cale)\b',
    ).allMatches(value).length;
    if (actionCount > 1 && RegExp(r'\b(?:et|puis|ensuite)\b').hasMatch(value)) {
      ambiguities.add('multiple_actions');
    }
    if (_hasNegation(value)) codes.add('negation_preserved');

    return NaturalLanguageNormalization(
      originalText: input,
      normalizedText: value,
      tokens: List.unmodifiable(
        value.isEmpty ? const <String>[] : value.split(' '),
      ),
      detectedLanguage: _detectLanguage(value),
      normalizationCodes: List.unmodifiable(codes),
      preservedAmbiguities: List.unmodifiable(ambiguities),
    );
  }

  static bool hasNegation(String input) => _hasNegation(
      const NaturalLanguageNormalizer().normalize(input).normalizedText);

  static bool _hasNegation(String value) => RegExp(
        r'\b(?:ne|n|pas|jamais|aucun|aucune|rien|personne|plus)\b',
      ).hasMatch(value);

  static String _detectLanguage(String value) {
    if (value.isEmpty) return 'und';
    final frenchMarkers = RegExp(
      r'\b(?:je|tu|il|elle|nous|vous|les|des|une|pour|demain|rendez|tache)\b',
    ).allMatches(value).length;
    return frenchMarkers > 0 ? 'fr' : 'und';
  }

  static String _foldAccents(String input) {
    const replacements = {
      'à': 'a',
      'â': 'a',
      'ä': 'a',
      'á': 'a',
      'ç': 'c',
      'é': 'e',
      'è': 'e',
      'ê': 'e',
      'ë': 'e',
      'î': 'i',
      'ï': 'i',
      'í': 'i',
      'ô': 'o',
      'ö': 'o',
      'ó': 'o',
      'ù': 'u',
      'û': 'u',
      'ü': 'u',
      'ú': 'u',
      'ÿ': 'y',
      'œ': 'oe',
    };
    var result = input;
    replacements.forEach((source, target) {
      result = result.replaceAll(source, target);
    });
    return result;
  }
}

File: functions/brain/engines/naturalLanguageNormalizer.js
"use strict";

const SAFE_TOKEN_REPLACEMENTS = Object.freeze({
  prebdre: "prendre",
  ajoutr: "ajouter",
  demian: "demain",
  deamin: "demain",
  deman: "demain",
  dem1: "demain",
  dmain: "demain",
  apresdemain: "apres demain",
  mtn: "maintenant",
  rapel: "rappel",
  rdv: "rendez vous",
  rendezvous: "rendez vous",
  rendezvou: "rendez vous",
  rendevous: "rendez vous",
  rendevouz: "rendez vous",
  rendezvouz: "rendez vous",
  crenau: "creneau",
  crenaux: "creneau",
  creneux: "creneau",
  heur: "heure",
  heurs: "heure",
  hure: "heure",
  hures: "heure",
  minut: "minute",
  mins: "minutes",
  stp: "s il te plait",
  svp: "s il vous plait",
  aujourdhui: "aujourd hui",
  auj: "aujourd hui",
  ajd: "aujourd hui",
  ojd: "aujourd hui",
  aprem: "apres midi",
  proposemoi: "propose moi",
  trouvemoi: "trouve moi",
  cherchemoi: "cherche moi",
  calemoi: "cale moi",
  ajoutemoi: "ajoute moi",
  rappellemoi: "rappelle moi",
  jveux: "je veux",
  jvoudrais: "je voudrais",
  doeuf: "d oeuf",
  doeufs: "d oeufs",
});

const UNDERSTANDING_LEVELS = Object.freeze([
  "exactMatch",
  "normalizedMatch",
  "probableMatch",
  "ambiguous",
  "noMatch",
]);
const NEGATED_ACTION = new RegExp(
    String.raw`\b(?:ne|n)\s+(?:cree|creer|ajoute|ajouter|annule|annuler|` +
    String.raw`supprime|supprimer|decale|deplacer|veux)\b.{0,40}` +
    String.raw`\b(?:pas|plus)\b`,
);
const ACTION_TOKEN = new RegExp(
    String.raw`\b(?:cree|ajoute|achete|annule|decale|deplace|supprime|` +
    String.raw`rappelle|planifie|cale)\b`,
    "g",
);

/**
 * Applies only bounded transformations that do not change user intent.
 *
 * @param {unknown} input source text
 * @return {{
 *  originalText: string,
 *  normalizedText: string,
 *  tokens: string[],
 *  detectedLanguage: string,
 *  normalizationCodes: string[],
 *  preservedAmbiguities: string[]
 * }}
 */
function normalizeNaturalLanguage(input) {
  const originalText = String(input || "");
  const normalizationCodes = [];
  const preservedAmbiguities = [];
  let value = originalText.trim();
  if (value !== originalText) normalizationCodes.push("trimmed");
  const lower = value.toLowerCase();
  if (lower !== value) normalizationCodes.push("lowercased");
  value = lower.replace(/[’‘`´]/g, "'");
  if (value !== lower) normalizationCodes.push("apostrophe_normalized");
  const folded = value.normalize("NFD").replace(/[\u0300-\u036f]/g, "")
      .replace(/œ/g, "oe");
  if (folded !== value) normalizationCodes.push("accents_folded");
  value = folded
      .replace(/['-]/g, " ")
      // A colon is meaningful inside a clock expression such as 14:30.
      .replace(/[.!?;,()[\]{}"…]+/g, " ")
      .replace(/[\u{1F300}-\u{1FAFF}]/gu, " ")
      .replace(/\s+/g, " ")
      .trim();

  const beforeOral = value;
  value = value
      .replace(/^jai\b/, "j ai")
      .replace(/^ya\b/, "y a")
      .replace(/^jveu\b/, "je veux")
      .replace(/^je veu\b/, "je veux")
      .replace(/\bneuf heure et demie\b/g, "09:30")
      .replace(/\bneuf heures trente\b/g, "09:30")
      .replace(/\bneuf heure\b/g, "09:00");
  if (value !== beforeOral) {
    normalizationCodes.push("oral_form_normalized");
  }

  let corrected = false;
  const tokens = value ? value.split(" ") : [];
  const correctedTokens = [];
  for (const token of tokens) {
    const replacement = SAFE_TOKEN_REPLACEMENTS[token];
    if (replacement) {
      correctedTokens.push(...replacement.split(" "));
      corrected = true;
    } else {
      correctedTokens.push(token);
    }
  }
  value = correctedTokens.join(" ");
  if (corrected) normalizationCodes.push("safe_typo_corrected");

  const beforePhrases = value;
  value = value
      .replace(
          new RegExp(
              String.raw`\b(?:la\s+)?(?:sem|semain|semiane|semaine)\s+` +
              String.raw`(?:proch|prochane|prochiane|prochaine)\b`,
              "g",
          ),
          "la semaine prochaine",
      )
      .replace(/\s+/g, " ")
      .trim();
  if (value !== beforePhrases) {
    normalizationCodes.push("safe_phrase_normalized");
  }

  if (/\b(?:je veux|ajoute)\s+plus\b/.test(value)) {
    preservedAmbiguities.push("positive_or_quantity_plus");
  }
  if (/\b(?:j ai|y a|on a)\s+plus\b/.test(value) &&
      !/\b(?:ne|n)\b/.test(value)) {
    preservedAmbiguities.push("stockout_or_additional_plus");
  }
  if (NEGATED_ACTION.test(value) ||
      /\b(?:cree|ajoute|annule|supprime|decale|veux)\s+pas\b/.test(value)) {
    preservedAmbiguities.push("critical_negation_scope");
  }
  if (/^(?:mets|met|decale)\s+(?:le|la|l)\b/.test(value)) {
    preservedAmbiguities.push("unresolved_reference");
  }
  const actionCount = (value.match(ACTION_TOKEN) || []).length;
  if (actionCount > 1 && /\b(?:et|puis|ensuite)\b/.test(value)) {
    preservedAmbiguities.push("multiple_actions");
  }
  if (hasNegationNormalized(value)) {
    normalizationCodes.push("negation_preserved");
  }

  return Object.freeze({
    originalText,
    normalizedText: value,
    tokens: Object.freeze(value ? value.split(" ") : []),
    detectedLanguage: detectLanguage(value),
    normalizationCodes: Object.freeze(normalizationCodes),
    preservedAmbiguities: Object.freeze(preservedAmbiguities),
  });
}

/**
 * @param {string} normalized normalized text
 * @return {boolean}
 */
function hasNegationNormalized(normalized) {
  return /\b(?:ne|n|pas|jamais|aucun|aucune|rien|personne|plus)\b/
      .test(normalized);
}

/**
 * @param {string} value normalized text
 * @return {string}
 */
function detectLanguage(value) {
  if (!value) return "und";
  const matches = value.match(
      /\b(?:je|tu|il|elle|nous|vous|les|des|une|pour|demain|rendez|tache)\b/g,
  );
  return matches && matches.length > 0 ? "fr" : "und";
}

module.exports = {
  UNDERSTANDING_LEVELS,
  normalizeNaturalLanguage,
  hasNegationNormalized,
};
